### PR TITLE
Added xla friendly codepath to single_tensor_adamw

### PR DESCRIPTION
**Pull request by aws-murandoo**
There are extra graph compilations on XLA when beta{1,2} ** step get too small. This PR addresses this issue by making the `capturable` interface enabled for XLA, as well as switching to `torch.float_power` which preserves the same behaviour as the non-capturable flow on XLA.

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
hmm, this looks correct to me, but I hesitate - do we need to have a more general way to update multiple optimizers for multiple device types? It won't be great to have this kind of change sprinkled everywhere.

**Comment by aws-murandoo** [comment]
Hey @wconstab I agree that there should be a better way to handle it in general, this approach is not as sustainable. However, considering the code is already CUDA-specific in places would it make sense to merge this change until a more general system is in place?

**Comment by wconstab** [comment]
Yea, i would defer to @albanD and @janeyx99 with the caveat that I don't think we should block this change, so if we can't propose a reasonable 'general way' quickly we should just merge this and clean up later.

File: torch/optim/adamw.py
Comment (albanD) [inline torch/optim/adamw.py:403]: This looks very surprising as it is a semantic change to what the operator does. This means that the correction will always be computed in float64 on xla device? Note that this will also impact all the other ops below and turn all of them in float64 ops!

Note that the non-capturable flow is not using float_power but `**` and so will preserve the dtype of the output contrary to this change.

Could you give more details why pow is different from `**` for you?

**Review by albanD — COMMENTED** [review]
This change does not look semantic preserving or matching the non-capturable path.
Left questions below

File: torch/optim/adamw.py
Comment (aws-murandoo) [reply torch/optim/adamw.py:403]: Sure! It is *not* the case that `float_power` computation will be done in float64 on the xla device. Internally, it will be lowered to an fp32 computation. 

`%power = f32[] power(f32[] %p1, f32[] %convert), metadata={op_type="aten__pow" op_name="aten__pow"}`

Results for the different ops:
`pow - 1.`
`float_power - 0.994`
`** - 0.994`

Test code for reference:
```
import torch
import torch_xla.core.xla_model as xm

beta = 0.995
def func1(step):
    step = torch.tensor(step, device=xm.xla_device())
    factor = 1 - torch.pow(beta, step)
    print(factor)

def func2(step):
    step = torch.tensor(step, device=xm.xla_device())
    factor = 1 - torch.float_power(beta, step)
    print(factor)

def func3(step):
    factor = 1 - (beta ** step)
    print(factor)

func1(100)
func2(100)
func3(100)
```

Comment (albanD) [reply torch/optim/adamw.py:403]: Interesting!
I'm wondering if we could move this back to just be `1 - beta1 ** step` in all cases then. Can you try to see if that passes CI please?

**Review by janeyx99 — COMMENTED** [review]
A general q: is this change only for capturable in adamw or is it also applicable to capturable in adam? I'd vouch for consistency with support there for XLA as well.

**Comment by aws-murandoo** [comment]
@janeyx99 I took a quick look, the same change looks like it'd be applicable to adam as well. Would you want me to fold that change into this PR as well?

**Comment by janeyx99** [comment]
Yes, and for common logic to be abstracted to optimizer.py

**Comment by aws-murandoo** [comment]
Sure, looks like the only additional change would be to adam.py, doesn't look like anything would be abstracted to optimizer.py needs currently.
Will go ahead once the discussion above with @albanD is cleared.

File: torch/optim/adamw.py
Comment (aws-murandoo) [reply torch/optim/adamw.py:403]: On further reflection `pow` would cause all optimizer computations to be in bf16 when using `XLA_USE_BF16` or `XLA_DOWNCAST_BF16`. Maybe `float_power` is the way to go.

Pinging @wconstab and @JackCaoG for comments.

**Comment by albanD** [comment]
Hey @aws-murandoo can you confirm the status on this? The pow operator has been updated and should now work fine.

**Comment by aws-murandoo** [comment]
Hey @albanD I am blocked on two PRs I've put out to PyTorch XLA. pow is still broken and giving incorrect results on XLA for certain numbers. Hopefully I can sync with XLA people on this. @lsy

pow: https://github.com/pytorch/xla/issues/5118
RFC: https://github.com/pytorch/xla/issues/5270

**Comment by albanD** [comment]
But I updated the adamw code to use the pure python operator. So the issues with `torch.pow` shouldn't block you here?

**Comment by aws-murandoo** [comment]
Ah yes I had to refresh my memory on https://github.com/pytorch/pytorch/pull/104254.

To summarize my understanding - in the capturable path we use torch.pow under the hood? I.e. `1 - beta1 ** step` is `1 - torch.pow(beta1, step)`.

I'll have a new revision up hopefully by today, should now be a 1-line change.

**Comment by aws-murandoo** [comment]
Hey @janeyx99 @albanD,

Just updated if you wanna take a look again.

**Comment by aws-murandoo** [comment]
Hey @albanD @janeyx99 

Did you get a chance to take a look at this?

**Review by janeyx99 — COMMENTED** [review]
There is now also an nadam capturable impl you may want to change as well.

**Comment by aws-murandoo** [comment]
I've added that as well right now @janeyx99

**Review by janeyx99 — APPROVED** [review]

**Comment by aws-murandoo** [comment]
Hi @albanD 
Just wanted to follow up on this one!

**Review by albanD — APPROVED** [review]
Ho yeah that's good for me!
Sorry I synced offline with @janeyx99 and didn't re-accept myself. But if @janeyx99 is happy, I'm happy!

**Comment by aws-murandoo** [comment]
Awesome!!

Let's merge! Thanks @janeyx99 @albanD 

FYI:
@wconstab @JackCaoG

**Pull request closed without merging**
